tensorflow/compiler/tf2tensorrt/convert/ops/layer_utils.h

```diff
@@ -0,0 +1,351 @@
+/* Copyright 2021 The TensorFlow Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+#ifndef TENSORFLOW_COMPILER_TF2TENSORRT_CONVERT_OPS_LAYER_UTILS_H_
+#define TENSORFLOW_COMPILER_TF2TENSORRT_CONVERT_OPS_LAYER_UTILS_H_
+#if GOOGLE_CUDA && GOOGLE_TENSORRT
+
+#include <type_traits>
+
+#include "tensorflow/compiler/tf2tensorrt/convert/convert_nodes.h"
+#include "tensorflow/compiler/tf2tensorrt/convert/utils.h"
+#include "tensorflow/core/framework/types.h"
+#include "tensorflow/core/platform/statusor.h"
+#include "third_party/tensorrt/NvInfer.h"
+#include "third_party/tensorrt/NvInferRuntimeCommon.h"
+
+namespace tensorflow {
+namespace tensorrt {
+
+namespace convert {
+
+class NetworkFactoryContext {
+ public:
+  NetworkFactoryContext(nvinfer1::INetworkDefinition* network,
+                        TrtWeightStore* weight_store)
+      : network_(network), weight_store_(weight_store) {}
+
+  template <typename T, typename std::enable_if<std::is_base_of<
+                            nvinfer1::ILayer, T>::value>::type* = nullptr>
+  struct CreationResult {
+    T* layer;
+    nvinfer1::ITensor* output;
+  };
```

Comment:
So most of the time, the functions in this class return a layer and the output of that layer, which two exception AbsInt and AbsCeilDivInt?

Here is a suggestion, to simplify the design and the interface of functions in this class:

  The "layer builder function" should return one value of the same type, either the layer itself or the output(0) of the layer, I will let you decide which value to return.
  For cases like AbsInt and AbsCeilDivInt, the users can call the layer builder multiple types.

Reply:
I definitely tried that. I agree a "single return type" is a nice interface feature. However, when you return only one, it is very painful either way. 

Returning only the layer means you need to add (`layer->getOutput(0)`) everywhere, as you need the `ITensor*` to pass to the next layer. Regardless, you've stated "one value of the same type" as a criterion, and while all the layers are derived from "ILayer", they are different types. Returning as ILayer pointless isn't really that useful. 

The above paragraph rules out returning the layer type. So what about just returning ITensor? That would be my choice. But unfortunately that doesn't work out either, because you often need to have the ILayer* pointer handy after creation (for example, for strided slice we use `layer->setInput(..)`. Futhermore, the existing funcitons in the `Converter` class like `setLayerName` require an `ILayer` pointer. Ideally we could move the name as a string parameter pushed forward to creation, similar to the TF C++ API, but that doesn't resolve point #1. 

I thought about working around this by adding a `getLastCreatedLayer`function to the builder. At first, this seems like a nice solution, since you only sometimes need the created layer, and you can just store a pointer to the last layer created. However, you also need the type of the last layer created, otherwise the user needs to specify the type, as in `auto layer = builder->getLastCreatedLayer<ISliceLayer>()`. So now we need a variable to keep track of the type of the last layer as well as some giant switch functor between enums and ILayer subtypes, and so on....

Now might see how my solution is actually pretty elegant, requiring minimal code, and providing the user everything they need.

Reply:
I'm only using Layer output a couple times in this function. But if you use the class elsewhere, you may need it in other places. For example, what if users want to apply some naming scheme to all the layers they create? In TensorRT you need the layer often.

Why does it bother to return struct? It is not inefficient, as the two fields are just pointers. What about returning std::pair?
Or I can put layer as an optional ptr parameter in each function. For uniformity and ease of use, I would like to make all the functions the same.

My preference is to keep as is, but if you still feel strongly about it, I can move the layer output to optional pointer parameter.

Reply:
Coding style prefer output returning values over output via pointer parameters. If "output" is always "output 0 from layer", then we should just return the layer instead of a pair of (output, layer). This is because "output" is redundant and can cause consistency issue (that is, how can we enforce in (output, layer) output = layer->output(0)?).

Reply:
I've applied your recommendation on just returning the layer.